src/CloudEvent.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+/**
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\CloudFunctions;
+
+class CloudEvent
+{
+    // Required Fields
+    private $id;
+    private $source;
+    private $specversion;
+    private $type;
+
+    // Optional Fields
+    private $datacontenttype;
+    private $dataschema;
+    private $subject;
+    private $time;
+    private $data;
+
+    public function __construct(
+        string $id,
+        string $source,
+        string $specversion,
+        string $type,
+        ?string $datacontenttype,
+        ?string $dataschema,
+        ?string $subject,
+        ?string $time,
+        ?array $data
+    ) {
+        $this->id = $id;
+        $this->source = $source;
+        $this->specversion = $specversion;
+        $this->type = $type;
+        $this->datacontenttype = $datacontenttype;
+        $this->dataschema = $dataschema;
+        $this->subject = $subject;
+        $this->time = $time;
+        $this->data = $data;
+    }
+
+    public function getId(): string
+    {
+        return $this->id;
+    }
+    public function getSource(): string
+    {
+        return $this->source;
+    }
+    public function getSpecVersion(): string
+    {
+        return $this->specversion;
+    }
+    public function getType(): string
+    {
+        return $this->type;
+    }
+    public function getDataContentType(): ?string
+    {
+        return $this->datacontenttype;
+    }
+    public function getDataSchema(): ?string
+    {
+        return $this->dataschema;
+    }
+    public function getSubject(): ?string
+    {
+        return $this->subject;
+    }
+    public function getTime(): ?string
+    {
+        return $this->time;
+    }
+    public function getData(): ?array
+    {
+        return $this->data;
+    }
+
+    public static function fromArray(array $arr)
+    {
+        $argKeys = ['id', 'source', 'specversion', 'type', 'datacontenttype', 'dataschema', 'subject', 'time', 'data'];
+        $args = [];
+        foreach ($argKeys as $key) {
+            $args[] = $arr[$key] ?? null;
+        }
+
+        return new static(...$args);
+    }
+
+    public function __toString()
+    {
+        $data_as_json = json_encode($this->data);
```

Comment:
Follow-up: Assumes data is JSON.

In the event of a false return value, we should fallback to including the data as-is. https://www.php.net/json_encode

For simplicity, we might also include the data unaltered since it may just as well be binary.